librosa/core/audio.py

```diff
@@ -208,6 +210,167 @@ def __audioread_load(path, offset, duration, dtype):
     return y, sr_native
 
 
+def stream(path, block_length, frame_length=2048, hop_length=512,
+           mono=True, offset=0.0, duration=None, fill_value=None,
+           dtype=np.float32):
+    '''Stream audio in fixed-length buffers.
```

Comment:
semantic nit, but maybe important ... should `block_length` be `buffer_length`? I was reading through to see what block_length controlled, figured out it was a frame buffer, and then saw the doc string header (here) called it fixed-length buffer, which grok'ed quickly.

Reply:
this also relates to some q's i've had / thrown around regarding buffering (stream-buffering) time-frequency matrices, where `librosa.util.frame` doesn't operate on ndarrays.rank > 1, but is _also_ a stream / buffer operation 🤔 ... I don't know what the consistent concept is here, but it feels like there's something to converge on?

Reply:
I would interpret "buffer length" as the number of samples, which is not exactly right here.

I think a better alternative might be `n_frames`?

Reply:
Aside: the reason that I didn't go with `n_frames` to begin with is that the output is not indexed by frames.  I don't think that it should be (for both technical and aesthetic reasons), so I think it's best to stick to block_length.

Of course, the documentation should be expanded to make it totally clear what a "block" means in this context, and what sort of results the user should expect.

Reply:
the output is _not_ indexed by frames? as a noob, i would expect something like the following:

```
>>> stream, sr = librosa.stream(librosa.util.example_audio_file(), 256)
>>> set([block.shape for block in stream])
{(256, 2048)}
```
librosa is feature-major, right? or is that not true for audio? MFCCs come back `(20, time)` ... if nothing else, a shape tuple in the docstring will help dummies like me

tangentially related, does it make sense that `frame_length` and `hop_length` have default values? had to scroll back to infer my shape example

Reply:
Just to clarify, `block_length` is not `n_frames` because..? Is it because interleaved n-channel audio files would have `block_length = n_frames * n_channel`? When I skimmed the code and docstrings I probably thought the same as @ejhumphrey.

But for whatever reason, seems like it's totally fair (or even better) to a new concept (`block`). Probably the docstring could start with `"""0. What is Block?:  .."""`.

Reply:
> librosa is feature-major, right? or is that not true for audio?

The convention is that time(-like dimensions) are trailing.  So for audio, it's (samples,) or (channels, samples), and for features it's (features, frames).


> tangentially related, does it make sense that `frame_length` and `hop_length` have default values? had to scroll back to infer my shape example

That's a great question.  In general, things should be consistent across the API assuming a default sampling rate.  Of course, that assumption doesn't apply here, so there's a good case for making these required parameters.



> the output is _not_ indexed by frames?

No, the output is a slice of audio.  I get that this is a little counter-intuitive from a parameterization perspective, but it makes translating current code into the streaming setting much easier (check the examples in the docstring).



> Just to clarify, `block_length` is not `n_frames` because..? Is it because interleaved n-channel audio files would have `block_length = n_frames * n_channel`?

The way I see it, it's because the output is not indexed by frame.  block_length can be thought of as a signal duration measured in frame units.



> But for whatever reason, seems like it's totally fair (or even better) to a new concept (`block`). Probably the docstring could start with `"""0. What is Block?: .."""`.

Yes, I agree with that completely.